src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2304,14 +2333,189 @@ private void CheckIndexerSignatureCollisions(
+                                membersByName.Add(pair.Key, concatMembers([], extension, pair.Value, ref membersUnordered));
+                            }
+                        }
+                    }
+                }
+
+                return (membersByName, membersUnordered);
+            }
+
+            static ImmutableArray<Symbol> concatMembers(ImmutableArray<Symbol> existingMembers, NamedTypeSymbol extension, ImmutableArray<Symbol> newMembers, ref ImmutableArray<Symbol> membersUnordered)
+            {
+                Debug.Assert(!newMembers.IsEmpty);
+
+                if (extension.IsDefinition)
+                {
+                    Debug.Assert(newMembers.All(static (m, membersUnordered) => membersUnordered.Contains(m), membersUnordered));
+                    return existingMembers.Concat(newMembers);
+                }
+
+                var membersBuilder = ArrayBuilder<Symbol>.GetInstance(existingMembers.Length + newMembers.Length);
+                var membersUnorderedBuilder = ArrayBuilder<Symbol>.GetInstance(membersUnordered.Length + newMembers.Length);
+
+                membersBuilder.AddRange(existingMembers);
+                membersUnorderedBuilder.AddRange(membersUnordered);
+
+                foreach (var member in newMembers)
+                {
+                    Symbol toAdd = member.SymbolAsMember(extension);
+                    membersBuilder.Add(toAdd);
+                    membersUnorderedBuilder.Add(toAdd);
+                }
+
+                membersUnordered = membersUnorderedBuilder.ToImmutableAndFree();
+                return membersBuilder.ToImmutableAndFree();
+            }
+        }
+
+        private readonly struct ExtensionGroupingKey : IEquatable<ExtensionGroupingKey>
+        {
+            public readonly int ExtensionArity;
+            public readonly TypeSymbol ReceiverType;
+
+            public ExtensionGroupingKey(NamedTypeSymbol extension)
+            {
+                ExtensionArity = extension.Arity;
+
+                if (extension.Arity != 0)
+                {
+                    extension = extension.Construct(IndexedTypeParameterSymbol.Take(extension.Arity));
+                }
+
+                if (extension.ExtensionParameter is { } receiverParameter)
+                {
+                    ReceiverType = receiverParameter.Type;
+                }
+                else
+                {
+                    ReceiverType = ErrorTypeSymbol.UnknownResultType;
+                }
+            }
+
+            public bool Equals(ExtensionGroupingKey other)
+            {
+                return ExtensionArity == other.ExtensionArity &&
+                       ReceiverType.Equals(other.ReceiverType, TypeCompareKind.AllIgnoreOptions);
+            }
+
+            public override bool Equals([NotNullWhen(true)] object? obj)
+            {
+                Debug.Assert(false); // Usage of this method is unexpected
+                return Equals((ExtensionGroupingKey)obj!);
```

Comment:
Conforming Equals override should handle `null`s, I think:
```suggestion
                return obj is ExtensionGroupingKey k && Equals(k);
```

Reply:
> Conforming Equals override should handle nulls, 

In general, this is true. However we control usage of this type and don't want it to be boxed for the purpose of comparison. Therefore, we not only do not expect it to be called with `null`, we do not expected it to be called ever. Hence the assert above.